ouster-ros/src/lidar_packet_handler.h

```diff
@@ -0,0 +1,288 @@
+        assert(scan_width + curr_scan_first_nonzero_idx >
+               last_scan_last_nonzero_idx);
+        double interpolated_value = linear_interpolate(
+            last_scan_last_nonzero_idx, last_scan_last_nonzero_value,
+            scan_width + curr_scan_first_nonzero_idx,
+            curr_scan_first_nonzero_value, scan_width);
+        return ulround(interpolated_value);
+    }
+
+    uint64_t extrapolate_value(int curr_scan_first_nonzero_idx,
+                               uint64_t curr_scan_first_nonzero_value) {
+        double extrapolated_value =
+            curr_scan_first_nonzero_value -
+            scan_col_ts_spacing_ns * curr_scan_first_nonzero_idx;
+        return ulround(extrapolated_value);
+    }
+
+    // compute_scan_ts_0 for first scan
+    uint64_t compute_scan_ts_0(
+        const ouster::LidarScan::Header<uint64_t>& ts_v) {
+        auto idx = std::find_if(ts_v.data(), ts_v.data() + ts_v.size(),
+                                [](uint64_t h) { return h != 0; });
+        assert(idx != ts_v.data() + ts_v.size());  // should never happen
+        int curr_scan_first_nonzero_idx = idx - ts_v.data();
+        uint64_t curr_scan_first_nonzero_value = *idx;
+
+        uint64_t scan_ns =
+            curr_scan_first_nonzero_idx == 0
+                ? curr_scan_first_nonzero_value
+                : extrapolate_value(curr_scan_first_nonzero_idx,
+                                    curr_scan_first_nonzero_value);
+
+        last_scan_last_nonzero_idx =
+            find_if_reverse(ts_v, [](uint64_t h) { return h != 0; });
+        assert(last_scan_last_nonzero_idx >= 0);  // should never happen
+        last_scan_last_nonzero_value = ts_v(last_scan_last_nonzero_idx);
+        compute_scan_ts = [this](const auto& ts_v) {
+            return compute_scan_ts_n(ts_v);
+        };
+
+        return scan_ns;
+    }
+
+    // compute_scan_ts_n applied to all subsequent scans except first one
+    uint64_t compute_scan_ts_n(
+        const ouster::LidarScan::Header<uint64_t>& ts_v) {
+        auto idx = std::find_if(ts_v.data(), ts_v.data() + ts_v.size(),
+                                [](uint64_t h) { return h != 0; });
+        assert(idx != ts_v.data() + ts_v.size());  // should never happen
+        int curr_scan_first_nonzero_idx = idx - ts_v.data();
+        uint64_t curr_scan_first_nonzero_value = *idx;
+
+        uint64_t scan_ns = curr_scan_first_nonzero_idx == 0
+                               ? curr_scan_first_nonzero_value
+                               : impute_value(last_scan_last_nonzero_idx,
+                                              last_scan_last_nonzero_value,
+                                              curr_scan_first_nonzero_idx,
+                                              curr_scan_first_nonzero_value,
+                                              static_cast<int>(ts_v.size()));
+
+        last_scan_last_nonzero_idx =
+            find_if_reverse(ts_v, [](uint64_t h) { return h != 0; });
```

Comment:
[nit]
works the same way as
```
std::find_if(std::make_reverse_iterator(ts_v.data() + ts_v.size()),
             std::make_reverse_iterator(ts_v.data()),
             [](auto&&){ .... });
```

Reply:
This so valid and it was what i initially attempted but it did fail to compile when using gcc-9.

Reply:
_Interesting_. Okay.